pkg/track/track.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package track
+
+import (
+	"strings"
+	"time"
+)
+
+type TrackTime struct {
+	P     time.Time
+	Costs []time.Duration
+}
+
+type Tracks struct {
+	datas    [MaxTrackPhase]TrackTime
+	times    [MaxTimestampPhase]time.Time
+	disabled bool
+}
+
+func (t *Tracks) Begin() {
+	if t.disabled {
+		return
+	}
+	if t.times[TrackStartTimestamp].IsZero() {
+		t.times[TrackStartTimestamp] = time.Now()
+	}
+}
+
+func (t *Tracks) StartTrack(phase TrackPhase) {
+	if t.disabled || phase >= MaxTrackPhase || phase <= NoTrack {
+		return
+	}
+	tk := t.datas[phase]
+	tk.P = time.Now()
```

Comment:
It can be optimized by `syscall.Gettimeofday`. but actual effect may need to be benchmark.

FYI:
https://github.com/go-eden/etime/blob/master/etime_now.go#L42

```
In my benchmark, the performance of etime.CurrentSecond was about 40 ns/op, the performance of time.Now() was about 68 ns/op:

BenchmarkCurrentSecond-12         	30000000	        40.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkCurrentMillisecond-12    	50000000	        39.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkCurrentMicrosecond-12    	30000000	        39.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkNowSecond-12             	20000000	        67.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkNowMillisecond-12        	20000000	        68.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkNowMicrosecond-12        	20000000	        67.8 ns/op	       0 B/op	       0 allocs/
```

Reply:
这个是两年以前的benchmark比较了，现在time.Now已经优化很多了。 另外syscall 对系统兼容性问题可能存在影响，还是不要用了